crates/core/tedge_api/src/message_log.rs

```diff
@@ -0,0 +1,152 @@
+//! The message log is a persistent append-only log of MQTT messages.
+//! Each line is the JSON representation of that MQTT message.
+//! The underlying file is a JSON lines file.
+use mqtt_channel::Message as MqttMessage;
+use serde_json::json;
+use std::fs::File;
+use std::fs::OpenOptions;
+use std::io::BufRead;
+use std::io::BufReader;
+use std::io::BufWriter;
+use std::io::Write;
+use std::path::Path;
+
+const LOG_FILE_NAME: &str = "entity_store.jsonl";
+
+#[derive(thiserror::Error, Debug)]
+pub enum LogEntryError {
+    #[error(transparent)]
+    FromStdIo(std::io::Error),
+
+    #[error("Deserialization failed with {0} while parsing {1}")]
+    FromSerdeJson(#[source] serde_json::Error, String),
+}
+
+/// A reader to read the log file entries line by line
+pub struct MessageLogReader {
+    reader: BufReader<File>,
+}
+
+impl MessageLogReader {
+    pub fn new<P>(log_dir: P) -> Result<MessageLogReader, std::io::Error>
+    where
+        P: AsRef<Path>,
+    {
+        let file = OpenOptions::new()
+            .read(true)
+            .open(log_dir.as_ref().join(LOG_FILE_NAME))?;
+        let mut reader = BufReader::new(file);
+
+        let mut version_info = String::new();
+        reader.read_line(&mut version_info)?;
+        // TODO: Validate if the read version is supported
+
+        Ok(MessageLogReader { reader })
+    }
+
+    /// Return the next MQTT message from the log
+    /// The reads start from the beginning of the file
+    /// and each read advances the file pointer to the next line
+    pub fn next_message(&mut self) -> Result<Option<MqttMessage>, LogEntryError> {
+        let mut buffer = String::new();
+        match self.reader.read_line(&mut buffer) {
+            Ok(bytes_read) if bytes_read > 0 => {
+                let message: MqttMessage = serde_json::from_str(&buffer)
+                    .map_err(|err| LogEntryError::FromSerdeJson(err, buffer))?;
+                Ok(Some(message))
+            }
+            Ok(_) => Ok(None), // EOF
+            Err(err) => Err(LogEntryError::FromStdIo(err)),
+        }
+    }
+}
+
+/// A writer to append new MQTT messages to the end of the log
+pub struct MessageLogWriter {
+    writer: BufWriter<File>,
+}
+
+impl MessageLogWriter {
+    pub fn new<P>(log_dir: P) -> Result<MessageLogWriter, std::io::Error>
+    where
+        P: AsRef<Path>,
+    {
+        let file = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(log_dir.as_ref().join(LOG_FILE_NAME))?;
+
+        // If the file is empty append the version information as a header
+        let metadata = file.metadata()?;
+        let file_is_empty = metadata.len() == 0;
+
+        let mut writer = BufWriter::new(file);
+
+        if file_is_empty {
+            let version = env!("CARGO_PKG_VERSION");
```

Comment:
@reubenmiller suggested that we use an independent version scheme for this persistence store file format rather than using the tedge version itself, which will make the validations easier in future, even when the same file format is used across multiple releases. So, this will be version `1.0` of the schema. And it will only be updated when there's a breaking change to the schema.